### PR TITLE
Improved the type checking for uses of scan.

### PR DESCRIPTION
**Pull request by gnecula**
Previous way of checking was done after flattening and got
easily confused by tuples of different shapes, or None.

Relates to https://github.com/google/jax/issues/1534.

### CODE REVIEW

**Review by mattjj — APPROVED** [review]
LGTM! These checks are much better, and thanks for adding the tests for error messages.

File: jax/lax/lax_control_flow.py [hunk 1]
```diff
@@ -474,10 +476,16 @@ def scan(f, init, xs):
...
   jaxpr, consts, out_tree = _initial_style_jaxpr(f, in_tree, carry_avals + x_avals)
-  carry_avals_out, y_avals = split_list(jaxpr.out_avals, [num_carry])
-  if tuple(carry_avals_out) != carry_avals:
+
+  out_tree_avals = tree_unflatten(out_tree, jaxpr.out_avals)
+  init_tree_avals = tree_unflatten(init_tree, carry_avals)
+  if not isinstance(out_tree_avals, tuple) or len(out_tree_avals) != 2:
+    msg = "scan body output must be a pair, got {}."
+    raise TypeError(msg.format(out_tree_avals))
+  if out_tree_avals[0] != init_tree_avals:
```
Comment (mattjj) [inline jax/lax/lax_control_flow.py]: While our standard pytrees (`list`/`tuple`/`dict`) define equality, this line would also require user-defined pytree types to define equality. That may be a new constraint on pytrees. I'm fine with adding that constraint here, but an alternative would be to compare the treedefs and separately compare the flattened lists of avals.

Comment (gnecula) [reply jax/lax/lax_control_flow.py]: Good point, I had tried that but I was not able to extract the treedef for the first element of a pair given the treedef for the pair. I need this to check that the returned carry (first element) matches the initial carry.

Comment (mattjj) [reply jax/lax/lax_control_flow.py]: I haven't looked at the details but [`treedef_children`](https://github.com/google/jax/blob/5b724e0019243cc5ebb71d65963e1d829f65999e/jax/tree_util.py#L122) might help us here.

**Comment by shoyer** [comment]
does this actually fix the error reporting problem from #1534, or is it just related?

**Comment by gnecula** [comment]
I have changed the implementation to compare pytreedefs instead of pytrees themselves. It was a good opportunity to learn more about how pytrees work. 
I have also expanded the tests to show the complete actual error message.

**Comment by gnecula** [comment]
> does this actually fix the error reporting problem from #1534, or is it just related?

It does fix it; one of the test cases directly targets this bug

File: jax/lax/lax_control_flow.py (hunk 1 above)
Comment (gnecula) [reply jax/lax/lax_control_flow.py]: Implemented suggestion. Custom pytrees can still be tricky, if the custom flattener stores aux_data that does not support equality.

**Pull request merged**